File: 0022-generate-parentheses/0022-generate-parentheses.go
func solve(opened, closed, n int, current string, res *[]string) {
    // If we have placed the desired number of open and close parentheses
    if opened == n && closed == n {
        *res = append(*res, current) // Add the current combination to the result
        return
    }
    
    // If we can still place an open parenthesis
    if opened < n {
        solve(opened+1, closed, n, current+"(", res) // Place an open parenthesis
    }

    // If we can still place a close parenthesis (but only if it matches an open one)
    if closed < opened {
        solve(opened, closed+1, n, current+")", res) // Place a close parenthesis
    }
}

func generateParenthesis(n int) []string {
    var res []string
    solve(0, 0, n, "", &res) // Start with no parentheses placed
    return res
}

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    private void generate(int opened, int closed, int n, StringBuilder sb, List<String> res) {
        if(opened == n && closed == n) {
            res.add(sb.toString());
        }

        if(opened < n) {
            sb.append('(');
            generate(opened + 1, closed, n, sb, res);
            sb.deleteCharAt(sb.length() - 1); // backtrack
        }
        if(closed < opened) {
            sb.append(')');
            generate(opened, closed + 1, n, sb, res);
            sb.deleteCharAt(sb.length() - 1); // backtrack
        } 
    }

    public List<String> generateParenthesis(int n) {
        StringBuilder sb = new StringBuilder();
        List<String> res = new ArrayList<>();
        generate(0, 0, n, sb, res);
        return res;
    }
}